lsp-completion.el

```diff
@@ -0,0 +1,553 @@
+         (goto-char bounds-start)
+         (and (lsp-completion--looking-back-trigger-characterp trigger-chars) t))
+       :company-match #'lsp-completion--capf-company-match
+       :company-doc-buffer (-compose #'company-doc-buffer
+                                     #'lsp-completion--capf-get-documentation)
+       :exit-function
+       (-rpartial #'lsp-completion--capf-exit-fn trigger-chars)))))
+
+(defun lsp-completion--capf-exit-fn (candidate _status &optional trigger-chars)
+  "Exit function of `completion-at-point'.
+CANDIDATE is the selected completion item.
+Others: TRIGGER-CHARS"
+  (unwind-protect
+      (-let* (((&plist 'lsp-completion-item item
+                       'lsp-completion-start-point start-point
+                       'lsp-completion-markers markers
+                       'lsp-completion-prefix prefix)
+               (text-properties-at 0 candidate))
+              ((&CompletionItem :label :insert-text? :text-edit? :insert-text-format? :additional-text-edits?)
+               item))
+        (cond
+         (text-edit?
+          (apply #'delete-region markers)
+          (insert prefix)
+          (lsp--apply-text-edit text-edit?))
+         ((or insert-text? label)
+          (apply #'delete-region markers)
+          (insert prefix)
+          (delete-region start-point (point))
+          (insert (or insert-text? label))))
+
+        (when (eq insert-text-format? 2)
+          (let ((yas-indent-line (lsp--indent-snippets?)))
+            (yas-expand-snippet
+             (lsp--to-yasnippet-snippet (buffer-substring start-point (point)))
+             start-point
+             (point))))
+
+        (when lsp-completion-enable-additional-text-edit
+          (if (or (get-text-property 0 'lsp-completion-resolved candidate)
+                  additional-text-edits?)
+              (lsp--apply-text-edits additional-text-edits?)
+            (-let [(callback cleanup-fn) (lsp--create-apply-text-edits-handlers)]
+              (lsp--resolve-completion-async
+               item
+               (lambda (resolved-item)
+                 (funcall callback
+                          (lsp:completion-item-additional-text-edits? resolved-item)))
+               cleanup-fn))))
+
+        (when (and lsp-signature-auto-activate
+                   (lsp-feature? "textDocument/signatureHelp"))
+          (lsp-signature-activate))
+
+        (setq-local lsp-inhibit-lsp-hooks nil)
+
+        (when (lsp-completion--looking-back-trigger-characterp trigger-chars)
+          (setq this-command 'self-insert-command)))
+    (lsp-completion--capf-clear-cache)))
+
+(defun lsp-completion--setup-company ()
+  "Setup company-mode."
+  (cond
+   ((and (functionp 'company-lsp)
+         (not lsp-completion-prefer-capf))
+    (progn
+      (company-mode 1)
+      (add-to-list 'company-backends 'company-lsp)
+      (setq-local company-backends (remove 'company-capf company-backends))))
+
+   ((and (fboundp 'company-mode) lsp-completion-enable)
```

Comment:
These should belong to auto configure

Reply:
Except the `add-hook` stuff, I think the part of configuring `company-lsp` or `company-capf` should be moved to auto configure